### DISCUSSION TITLE
Importing is extraordinarily slow when finding duplicates

### DISCUSSION THREAD

**Discussion by veryheavypickle** [photoprism/photoprism#3775] [Q&A] [answered]
PLEASE PROCEED ONLY IF YOU ARE ABSOLUTELY SURE THAT THIS IS NOT A TECHNICAL SUPPORT INCIDENT AND/OR POSSIBLY A PROBLEM WITH SOME OTHER SOFTWARE YOU ARE USING. VISIT <https://www.photoprism.app/kb/getting-support> TO LEARN MORE ABOUT OUR SUPPORT OPTIONS. THANK YOU FOR YOUR CAREFUL CONSIDERATION!

---------------------------------------------------------------------------

#### 1. What is not working as documented?

When running an import from media already placed in the imports folder containing duplicated images, the import runs excruciatingly slow. About 50 images per hour (normal jpegs of around 2 mb in size). These are photos that are already on the machine where the docker compose is running. When importing via the upload button in the web from a different computer (even with duplicates), the import runs way faster, about 50 GB of photos uploaded per hour from start of upload to finishing indexing. This is what I had been doing originally however due to multiple folders I decided to place all folders in the import folder and just import.

Reasons for duplicates is due to having found photos recovered from a hard drive.

On average:
CPU doesn't go above 3%, ram usage is 2.5 GiB swap usage is 0 KiB.

#### 2. How can we reproduce it?

1. Add folder of media (that is already in the library) to /photorpism/import using scp or SFTP
2. In the photoprism web interface go to Library > Import
3. Make sure "Move Files" is selected
4. Watch a single photo of 2 megabytes get processed in a minute and a half (22 kilobytes/second)

See logs below

Library currently has 110,000 photos and the sum of all the folders of photos in import/ is nearly 200,000.

#### 3. What behavior do you expect?

I expect there to not be a delay of 60 to 100 seconds after finding a duplicate.

#### 4. What could be the cause of your problem?

I am using a NAS, however moving the hundreds of thousands of images from multiple sources to the import folder took a few hours.

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

```
started 230719 as uid 1000 (amd64-prod)
init: https
Creating a default HTTPS/TLS certificate.
Generating self-signed issuer (CA) certificate...
Generating self-signed tls certificate...
Certificate request self-signature ok
subject=C = DE, ST = Berlin, L = Berlin, O = PhotoPrism, OU = Self-Hosted, emailAddress = hello@photoprism.app, CN = PhotoPrism
Updating permissions of keys in '/etc/ssl/private'...
Running 'update-ca-certificates'...
Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
Problems? Our Troubleshooting Checklists help you quickly diagnose and solve them:
https://docs.photoprism.app/getting-started/troubleshooting/
file umask....: "0002" (u=rwx,g=rwx,o=rx)
home directory: /photoprism
assets path...: /opt/photoprism/assets
storage path..: /photoprism/storage
config path...: default
cache path....: default
backup path...: /photoprism/storage/backups
import path...: /photoprism/import
originals path: /photoprism/originals
running as uid 1000
/opt/photoprism/bin/photoprism start
time="2023-09-29T17:45:29Z" level=info msg="config: case-insensitive file system detected"
time="2023-09-29T17:45:29Z" level=info msg="Become a member today, support our mission and enjoy our member benefits! 💎"
time="2023-09-29T17:45:29Z" level=info msg="Visit https://www.photoprism.app/membership to learn more."
time="2023-09-29T17:45:29Z" level=info msg="migrate: running database migrations"
time="2023-09-29T17:45:31Z" level=info msg="server: enabled gzip compression"
time="2023-09-29T17:45:31Z" level=info msg="webdav: shared /originals/"
time="2023-09-29T17:45:31Z" level=info msg="webdav: shared /import/"
time="2023-09-29T17:45:31Z" level=info msg="server: disabled tls"
time="2023-09-29T17:45:31Z" level=info msg="server: listening on 0.0.0.0:2342 [1.732044ms]"
time="2023-09-29T17:45:42Z" level=info msg="moving files from import"
time="2023-09-29T17:45:42Z" level=info msg="classify: loading nasnet"
time="2023-09-29T17:45:44Z" level=info msg="classify: loading labels from labels.txt"
time="2023-09-29T17:45:44Z" level=info msg="import: found no .ppignore file"
time="2023-09-29T17:45:44Z" level=info msg="import: added folder /all-jpeg"
time="2023-09-29T17:45:49Z" level=info msg="media: 'Apple iPhone 4032x3024_162502.jpg' was taken at 2018-04-05 07:46:02 +0000 UTC (meta)"
time="2023-09-29T17:45:49Z" level=info msg="import: 'Apple iPhone 4032x3024_162502.jpg' is identical to 2018/04/20180405_074602_93F0E750.jpg (sha1 b1590b7ea605109e15799459dbbc9ecf6febc605)"
time="2023-09-29T17:45:49Z" level=info msg="import: deleted 'all-jpeg/Apple iPhone 4032x3024_162502.jpg' (already exists)"

time="2023-09-29T17:48:18Z" level=info msg="media: 'Apple iPhone 4032x3024_162507.jpg' was taken at 2018-04-05 08:10:02 +0000 UTC (meta)"
time="2023-09-29T17:48:18Z" level=info msg="import: 'Apple iPhone 4032x3024_162507.jpg' is identical to 2018/04/20180405_081002_F4E34FE8.jpg (sha1 6861939f32ef79280b5d405a85ce0efc0dd2de3a)"
time="2023-09-29T17:48:18Z" level=info msg="import: deleted 'all-jpeg/Apple iPhone 4032x3024_162507.jpg' (already exists)"
```
#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: amd64-prod, Build 230719-73fa7bbe8

(b) Database Type & Version: MariaDB

(c) Operating System Types & Versions: Ubuntu 22.04

(d) Browser Types & Versions: Brave/Chromium

(e) Ad Blockers, Browser Plugins, and/or Firewall Software? Pihole + Unbound and WireGuard

#### 7. On what kind of device is PhotoPrism installed?

(a) Device / Processor Type: Ubuntu VM in Proxmox (no gpu) AMD Ryzen 5 5600G

(b) Physical Memory & Swap Space in GB: 16 GiB RAM, 8 GiB Swap (dedicated ram/balloon)

(c) Storage Type: SMB Nas from TrueNAS, 5 x 8 TB WD UltraStar disks (2 are redundant)

(d) Anything else that might be helpful to know?
1. The host machine is currently dedicated to importing photoprism and has ram reserved for this machine.
2. Using photoprism is fast, scrolling through all photos is very fast, reading photos is fast, searching is fast, uploading/downloading is fast, renaming problematic (folders with duplicates) folders (so it gets imported later) and importing unimported files is fast.
3. The size of the library is about 2 TB.
4. The library is located on a NAS share (smb) that is mounted in the fstab - photoprism folder structure is default.

#### 8. Do you use a Reverse Proxy, Firewall, VPN, or CDN?

Yes, I use NGINX Proxy Manager with letsencrypt cert and a custom domain name that points to a local IP address.

**Comment by graciousgrey** [comment]
Thank you very much for your feedback!

I just tested this with a Docker Compose setup without SMB share and could not reproduce the delays you observed. The bottleneck in this case might be your setup/environment. In this case it might be better to do the import with the "Move" option enabled.

https://docs.photoprism.app/user-guide/library/import/#when-should-move-files-be-selected

Edit: I just noticed you do have the move option enabled. I will test that as well.

Importing (with move enabled) a 30GB import folder with files that already exist in originals took only a few seconds. There was no delay at all.

**Comment by veryheavypickle** [comment]
Hello!

While doing my own testing I realised that folders with smaller amounts of photos in them (and some duplicates) work very fast. However if it is a folder with (in my case) nearly 1 million photos, not separated by subfolders (my photo recovery program dumped them all into one giant folder) it is very slow.

I have two of these mega folders, one of them with 300,000 photos and the other with 900,000. When I created this bug report/discussion I was importing the 300,000 photo folder and there was a delay of 1-2 minutes. However in the 900,000 photo one, there is a delay of about 30 minutes between photo duplicate detection. The 900,000 folder is about 270GB

I haven't looked deep enough into the code and I don't know go, but this import delay is exponential, I haven't seen anything that would seem to cause this exponential gap in duplicate import times.

I guess my solution would be to separate the mega folders into subfolders of 1000 photos each. I will post an update :)

**Comment by veryheavypickle** [comment, marked as answer]
Solution found,

The issue is having tonnes of files in a single folder, running the small bash script below, [copied from stackoverflow](https://stackoverflow.com/questions/29116212/split-a-folder-into-multiple-subfolders-in-terminal-bash-script), I was able to reduce the import to a tens of images/second.

```
#!/bin/bash

dir_size=2000
dir_name="folder"
n=$((`find . -maxdepth 1 -type f | wc -l`/$dir_size+1))
for i in `seq 1 $n`;
do
    mkdir -p "$dir_name$i";
    find . -maxdepth 1 -type f | head -n $dir_size | xargs -i mv "{}" "$dir_name$i"
done
```

For the 900,000 file folder, this reduces the import time from about 50 years down to a couple of hours. For the script above, I left the default setting of 2000 files/folder.

**Reply by graciousgrey** [reply]
Glad to hear that and thank you for the update :)
